src/ImageSharp/Formats/ImageDecoder.cs

```diff
@@ -0,0 +1,246 @@
+    protected abstract IImageInfo Identify(DecoderOptions options, Stream stream, CancellationToken cancellationToken);
+
+    /// <summary>
+    /// Performs a scaling operation against the decoded image. If the target size is not set, or the image size
+    /// already matches the target size, the image is untouched.
+    /// </summary>
+    /// <param name="options">The decoder options.</param>
+    /// <param name="image">The decoded image.</param>
+    protected static void ScaleToTargetSize(DecoderOptions options, Image image)
+    {
+        if (ShouldResize(options, image))
+        {
+            ResizeOptions resizeOptions = new()
+            {
+                Size = options.TargetSize.Value,
+                Sampler = options.Sampler,
+                Mode = ResizeMode.Max
+            };
+
+            image.Mutate(x => x.Resize(resizeOptions));
+        }
+    }
+
+    /// <summary>
+    /// Determines whether the decoded image should be resized.
+    /// </summary>
+    /// <param name="options">The decoder options.</param>
+    /// <param name="image">The decoded image.</param>
+    /// <returns><see langword="true"/> if the image should be resized, otherwise; <see langword="false"/>.</returns>
+    private static bool ShouldResize(DecoderOptions options, Image image)
+    {
+        if (options.TargetSize is null)
+        {
+            return false;
+        }
+
+        Size targetSize = options.TargetSize.Value;
+        Size currentSize = image.Size();
+        return currentSize.Width != targetSize.Width && currentSize.Height != targetSize.Height;
+    }
+
+    internal static T WithSeekableStream<T>(
+        DecoderOptions options,
+        Stream stream,
+        Func<Stream, T> action)
+    {
+        Guard.NotNull(options, nameof(options));
+        Guard.NotNull(stream, nameof(stream));
+
+        if (!stream.CanRead)
+        {
+            throw new NotSupportedException("Cannot read from the stream.");
+        }
+
+        T Action(Stream s, long position)
+        {
+            T result = action(s);
+
+            // Issue #2259. Our buffered reads may have left the stream in an incorrect non-zero position.
+            // Reset the position of the seekable stream if we did not read to the end to allow additional reads.
+            if (stream.CanSeek && stream.Position != s.Position && s.Position != s.Length)
+            {
+                stream.Position = position + s.Position;
+            }
+
+            return result;
+        }
+
+        if (stream.CanSeek)
+        {
+            return Action(stream, stream.Position);
+        }
+
+        Configuration configuration = options.Configuration;
+        using ChunkedMemoryStream memoryStream = new(configuration.MemoryAllocator);
+        stream.CopyTo(memoryStream, configuration.StreamProcessingBufferSize);
+        memoryStream.Position = 0;
+
+        return Action(memoryStream, 0);
```

Comment:
It seems to have the same issue I mentioned in my previous review:
`stream.CanSeek` is always false on this branch, therefore this is equivalent of the following:

```suggestion
        return action(memoryStream, 0);
```

(I don't mean to apply this suggestion, better to inline `Action`.)

Reply:
Not sure I understand. We want to call Action and the code is called from multiple branches.

Reply:
After the code doesn't return here, `stream.CanSeek` will be always `false`:

https://github.com/SixLabors/ImageSharp/blob/8fdd6b0e6324e9ff405c4fc822d81ffe6cef051c/src/ImageSharp/Formats/ImageDecoder.cs#L171-L174

In that case `if (stream.CanSeek & ....` in `Action` doesn't make sense.

Reply:
[6978235](https://github.com/SixLabors/ImageSharp/pull/2276/commits/69782359ba4718d232c4cd62bc592e5065a7354e) removed the `stream.CanSeek` check in the sync path, effectively activating the rewinding logic, without any impact to our test result outcome. To me this is a sign that our functional coverage is incomplete.

Reply:
Ah yes of course! We cannot reset anyway since the entire input stream is copied. Will update.

Reply:
> [6978235](https://github.com/SixLabors/ImageSharp/pull/2276/commits/69782359ba4718d232c4cd62bc592e5065a7354e) removed the `stream.CanSeek` check in the sync path, effectively activating the rewinding logic, without any impact to our test result outcome. To me this is a sign that our functional coverage is incomplete.

That's not true. The CanSeek part was superfluous as the stream was always seekable at that point as you noted elsewhere.

Reply:
Ah, hold on. For sync, no, for async yes. Will add a non-seekable input stream to cover.

Reply:
Have updated the implementation and test both seekable and non-seekable scenarios